### PR TITLE
Use tau as a library for the tarski-bot

### PR DESCRIPTION
**Pull request by etorreborre**
@charleso this goes hand in hand with an upcoming PR for `tarski-bot`. Basically I dropped the `main` function and replaced it with a function accepting a different callback for each bot.

### CODE REVIEW

File: MainBot.hs [hunk 1]
```diff
@@ -19,24 +17,18 @@ import Data.XML.Types (
...
-  return ()
-
-main :: IO ()
-main = do
+-- | function used by a bot to answer request
+--   the general form is
+--     (configuration, from user, to user, message) -> IO ([messages], updated configuration)
+--
+--   the configuration can be updated as a way to keep some state between invocations
+type ReceiveMessage c = (c -> String -> String -> String -> IO ([String], c))
```
Comment (charleso) [inline MainBot.hs]: Probably worth adding at least a `User` `newtype` for to/from while we're here?

File: MainBot.hs [hunk 2]
```diff
@@ -1,7 +1,5 @@
 {-# LANGUAGE OverloadedStrings #-}
-module Main where
-
-import Bot
+module MainBot where
```
Comment (charleso) [inline MainBot.hs]: If we're going to do this, should probably live in `src/Tau.hs`

Comment (etorreborre) [inline MainBot.hs]: You mean just rename MainBot as Tau?

**Comment by charleso** [comment]
This stuff definitely needs some love, but I think this is a step in the right direction.

You mind adding a branch ci script while you're here? (Separate PR is fine too).

:+1:

File: MainBot.hs [hunk 3]
```diff
@@ -73,13 +65,29 @@ main = do
...
   -- FIX Without the fork tarksi-bot seems to ignore signals
-  _ <- forkIO $ forever (mainLoop conf sess)
+  _ <- forkIO $ forever (mainLoop conf receiveMessage sess)
 
   _ <- installHandler sigINT (Catch $ putMVar bye True) Nothing
   _ <- installHandler sigTERM (Catch $ putMVar bye True) Nothing
   _ <- takeMVar bye
   return ()
 
+mainLoop :: IORef c -> ReceiveMessage c -> Session -> IO ()
```
Comment (charleso) [inline MainBot.hs]: Food for thought - the `IORef` can also be rolled up in to the `IO` of `ReceiveMessage`. Some bots probably don't care about the state. You _could_ do the following:

```haskell
type ReceiveMessage = String -> String -> String -> IO [String]

receiverWithState :: IO c -> (c -> String -> String -> String -> IO ([String], c)) -> IO ReceiveMessage
```

Not recommending this over just leaving what's here, but just thought I'd mentioned it.

File: MainBot.hs (hunk 2 above)
Comment (charleso) [inline MainBot.hs]: > You mean just rename MainBot as Tau?

And put it in a new `src/` directory.

**Pull request merged**